---
title: 【JavaScript】오늘 새벽 무렵, [ async - await ]씨가 시체로 발견되어...
tags: JavaScript 비동기처리 promise async 레퍼
author: impl_s
slide: false
---

```javascript
// 모든 데이터와 오래된 데이터를 비교하고, 새 데이터만을 표시한다
async function showNewData() {
  const allData = await fetchAllData();
  const oldData = await fetchOldData();
  showData(allData, oldData);
}
```

![nanjyakorya.GIF](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/6a653e42-9914-6339-ee47-e57e95c863c1.gif)

**「뭐, 뭐야 이게에에에에에넴엚ㅁㄴㅇㄹㄴㅇㄹ！！！！」**


・・・

・・・

・・・


# 비동기 처리를 "제대로" 이해하고 사용하고 싶다

『비동기 처리를 어찌저찌 대충 이해해서 쓰고 있다...』

『일단 움직이니까, OK!』

**레이와 프로그래머***1인 제 자신이, 되도록 조심해서 쓰려고 하고 있습니다만, 저도 모르게 대충 쓰게 되는 경향이 있습니다.

（*1: 레이와 이후에 프로그래밍을 알게 된 사람. 초심자를 뜻함.）


참고로 위 코드는, **「병렬로 할 수 있는 처리를 직렬로 하고 있는」** 예시입니다. 


# 실제로 손으로 움직여보면서 비동기 처리를 깊이 이해해보자

본 기사는 아래의 구성으로 되어 있으며, 순서대로 비동기 처리를 학습하여 내일부터 자신을 가지고 비동기 처리를 쓸 수 있게 되기 위한 핸즈온입니다.


- JS의 비동기 처리에 대해 알기
- Promise에 대해 알기
- Async / Await를 사용할 수 있게 되기
- 윗 코드를 리팩터링 할 수 있게 되기

실제로 손으로 움직여볼 수도 있는 내용으로 되어 있습니다.

# 갑작스럽지만 문제입니다

아래 2개의 차이, 알 수 있나요?

```javascript
// ①
fetch("https://qiita.com/api/v2/items");
// ②
await fetch("https://qiita.com/api/v2/items");
```

게다가 아래 2개의 차이도 알 수 있나요?

```javascript:①
const func = () => {
  console.log('좋은 아침입니다');

  Promise.resolve('점심 맛있게 드세요')
    .then(res => console.log(res));

  console.log('좋은 밤 보내세요');
}
func();
```

```javascript:②
const func = async () => {
  console.log('좋은 아침입니다');

  await Promise.resolve('점심 맛있게 드세요')
    .then(res => console.log(res));

  console.log('좋은 밤 보내세요');
}
func();
```
이미 알고 있는 사람에게는, 이 기사는 장황한 설명이 될 수도 있습니다.

**分からない人なら、本記事を読むことで**
**JaveScriptの非同期処理の理解が深まり、上記２つの問題に答えることができるようになるかもしれません。**



# JS는 싱글 스레드

JS는 싱글 스레드입니다. 즉 **동시에 다룰 수 있는 작업은 1개 뿐**입니다.
프로그램을 순차적으로 1개씩 처리합니다.

![shopping_reji_gyouretsu_man.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/31ba1abe-6161-deee-c990-63ab124c7c94.png)


그렇지만, 그렇게 되면 곤란해지는 부분도 있습니다. 시간이 걸리는 처리가 도중에 생긴다면 어떻게 될까요? 예를 들면...

- 외부에 있는 DB로의 조회 작업
- 이미지 처리

이와 같은 작업은 완료하는데 얼마나 시간일 걸릴지 모릅니다.

순차적으로 하나씩 처리해가는 도중에, 갑자기 처리가 멈추고, 그 다음 언제 재개될지 모릅니다.

거기서, **『비동기』가 나설 차례**입니다.

시간이 걸리는 처리는 일단 두고, 금방 완료할 수 있는 다른 처리부터 차근차근 실행해 나갈 수 있습니다.


# Promise는 언젠가 값이 돌아온다는 약속
![yubikiri_business.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/75b4668f-fc53-05da-9f13-b81004b3ff0c.png)

아까 **「시간이 걸리는 처리는 일단 두고」**라고 말했지만, 제대로 그 처리 결과가 돌아올지 불안합니다.

거기서 등장하는 것이 **「Promise」** 입니다.

「DB에서 값을 가지고 올 테니까 기다려 줘」라고 약속하는 것입니다.


# 실제로 API에서 값을 가져와 보기

그러면, 실험 삼아 Qiita의 API에서 기사 데이터를 가져와 봅시다.

Chrome 등의 브라우저의 개발자 툴을 열고, 콘솔에서 아래 코드를 실행합니다.


```javascript
fetch("https://qiita.com/api/v2/items");
```

↓ 아래 사진은 실제로 브라우저의 콘솔에서 실행한 상태를 찍은 것입니다.
![スクリーンショット 2020-08-21 23.34.00.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/b3c37a6f-5490-672f-4661-fedb3daba356.png)


`Promise`가 표시되어 있습니다.
(그리고 `<pending>`라고도 표시되어 있습니다)

내용물을 보면 아래와 같이 표시되었습니다.
![スクリーンショット 2020-08-21 23.31.15.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/84806aeb-e888-6d82-decc-be71770350b1.png)

`Response`라는 것이 돌아온 모양입니다.

이건 Qiita의 기사를 가져오는 API이지만, 여기서는 아직 기사 목록은 볼 수 없는 것 같습니다.

그리고, **상태가 `fulfilled (약속이 지켜졌다)`로 되어 있습니다**

# Promise의 상태

`Promise`에는 `fulfilled`와 같은 스테이터스(상태)가 아래와 같이 3개 있습니다.

- `pending` - 보류 중
- `fulfilled` - 이행되었음
- `rejected` - 각하되었음

약속이 지켜지는 것을 기다리고 있음・약속이 지켜짐・약속이 깨짐

3가지 패턴입니다.


## 실제로 Promise를 만들어보기

아래 코드를 브라우저 콘솔에 입력하여 실제로 Promise를 만들어보는 것으로 더욱 더 이해할 수 있습니다.

### 계속 기다리게 되는 Promise

```javascript
const p1 = new Promise(() => {});
```
`Promise`의 생성자에는 콜백을 넘깁니다.

약속이 지켜졌을 때의 처리나, 지켜지지 못했을 때의 처리를 적기 위해서입니다.

이 `p1`이라고 이름 지은 `Promise`에는 아무것도 없는 콜백을 넘겼습니다.

![スクリーンショット 2020-08-21 23.52.09.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/bf6ad85e-b784-6205-73df-5be470e7ce71.png)

`p1`은 `Promise`고, `pending`(보류중) 상태입니다.

약속이 지켜지는 것을 기다리고 있지만, 콜백에는 아무 것도 적혀있지 않기 때문에 아무리 기다려도 보류 중인 상태입니다.


### 약속이 지켜지는 Promise

사실은, `Promise`의 콜백에는 2개의 파라미터를 설정할 수 있습니다.

- resolve
- reject

위의 2개입니다.

이번에는, 아래와 같이 브라우저 콘솔에 입력합니다.

```javascript
const p2 = new Promise((resolve, reject) => resolve('해결'));
```

コールバックの中で、`resolve()`を使用してみました。
![スクリーンショット 2020-08-21 23.53.23.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/879d8007-7f80-90ef-8853-d28c817a2a49.png)

`p2`의 상태는 `fulfilled`가 되어 있습니다.

그리고, `PromiseValue` 안에는 `resolve()`로 넘긴 파라미터가 들어 있습니다.


### 약속이 깨지는 Promise

그러면 이번에는, `reject()`를 콜백 내에서 사용해 봅시다.

```javascript
const p3 = new Promise((resolve, reject) => reject('기각'));
```
![スクリーンショット 2020-08-21 23.56.26.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/8065c05e-6294-a7e8-787f-130718bb544f.png)

`p3`에는 스테이터스가 `rejected`로 되어 있습니다.

**그리고, 콘솔에는 빨간색으로 오류 표시도 되어 있습니다!**

**『약속이 깨졌다』는건 오류**라는 뜻입니다.


# 여기까지 정리 ①

실제로 `Promise`를 작성해 봄으로써 아래 내용을 알 수 있었습니다.

- 비동기 처리의 리턴 값은 일단 `Promise`
- `Promise`에는 상태가 있다
- 무사히 값이 돌아왔다 -> `fulfilled`
- 오류가 돌아왔다 -> `rejected`




# then() 와 catch()

**`Promise`로 돌아온 값을 다루기 위해서는 `then()`와 `catch()`라는 2개의 메소드를 사용합니다.**（그 외에도 있으나 여기서는 다루지 않습니다.）

`then()`하고 `catch()`는 `Promise`에서 사용할 수 있는 메소드로,
**콜백 안에서 `PromiseValue`(즉 Promise에 들어온 값)**을 다룰 수 있습니다.

각각, 정상 처리와 이상 처리를 담당합니다.

- `then()`은 `resolve()`(해결)된 `PromiseValue`를 담당
- `catch()`은 `reject()`(각하)된 `PromiseValue`를 담당

그럼, 아까 만들었던 `Promise`에 대해 `then()`와 `catch()`를 사용해 보겠습니다.

### `then()`의 경우

```javascript
p2.then(res => console.log(res));
```

![スクリーンショット 2020-08-22 0.09.59.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/a3d66f62-bca8-ff85-6bed-30123d72aec5.png)

`PromiseValue`「해결」 문자열이 출력되어 있습니다.

또, 새로운 `Promise`가 리턴되어 있습니다.

### `catch()`의 경우

```javascript
p3.catch(err => console.log(err));
```

![スクリーンショット 2020-08-22 0.12.05.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/ba253e12-af55-f00a-befc-743723b9f784.png)

`PromiseValue`「각하」 문자열이 출력되어 있습니다.

또, 여기에도 새로운 `Promise`가 리턴되어 있습니다.


**이런 식으로, `then()`과 `catch()`를 사용함으로써 `Promise`의 결과에 접근할 수 있습니다.**

## Qiita 기사 API를 호출해서, `then()`으로 Promise 결과를 보기

그럼, 이번에는 Qiita API로 기사 목록을 받아서 열람해 봅시다.

아까 본 기사의 첫 부분에서 Qiita API를 fetch 했을 때는, `Promise`만 리턴되고 중요한 데이터는 볼 수 없었습니다.

이번에는, `then()`을 사용해서 Promise로 돌아온 값에 접근합니다.

```javascript
(fetch("https://qiita.com/api/v2/items")).then(res => console.log(res.json()));
```
![スクリーンショット 2020-08-22 1.50.53.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/b9c81716-133c-f179-6f6d-037153627e93.png)

기사 정보를 받아볼 수 있었습니다!


# then()을 연결하기

`then()`은 연결해서 적을 수도 있습니다.

아래 프로그램을 봐 주세요.

```javascript
getImage(file)
  .then(image => compressImage(image))
  .then(cImage => saveImage(cImage))
  .then(result => console.log(result))
  .catch(err => {throw new Error(err)})
```

`then()`이 잔뜩 나열되어 있습니다.

위 프로그램은 시간이 걸리는 사진 처리를 상상한 것입니다.

- 사진 파일을 읽어오기
- 사진을 읽어오면, 사진을 압축하기
- 사진이 압축되면, 그 사진을 DB에 저장하기
- 저장한 결과를 출력하기

ファイルの読み込みや、圧縮、保存、といった処理はそれぞれ時間がかかるので、**次の工程に進むためには、前の工程の完了を待つ必要があります。**

`Promise.then()`を使うことで、前の処理の結果を**待ち**、その結果を次の処理に渡すことが可能です。

こういう処理は、通常の同期処理ではうまくいかないので、`Promise`を用いた非同期処理が適しています。

また、最後の`catch()`はこのプログラムのどこでエラーが発生したとしても、うまくエラーをキャッチしてくれます。
なぜかというと、`then()`は`fulfilled`以外の場合は全てスルーして次の処理に渡すからです。

エラーの場合、ステータスは`rejected`になっているので、全ての`then()`はこれをスルーします。最後に残った`catch()`だけがこれを掴むことができます。


# ここまでのまとめ②

一休みして、ここまでのおさらいをして整理しておきます。

- JSはシングルスレッドであり、順番に同期処理がキホン
- 時間のかかる処理や、結果が不明な処理は非同期処理したい
- 非同期処理の結果を約束する`Promise`がある
- Promiseには非同期処理の結果が返ってくる
- `Promise`は`fulfill`されたり、`reject`されたりする
- `fulfill`された場合の値は、`then()`で取得できる
- `reject`された場合の値は、`catch()`で取得できる


# Promiseが処理されるタイミング

非同期処理を扱う際は、**実行タイミング**に気を配る必要があるでしょう。

以下のコードを見てください。単純なログを出力するプログラムですが、`console.log()`出力は『こんにちは』が一番最後になってしまいます。

```javascript:プログラム
const func = () => {
  console.log('おはよう');

  Promise.resolve('こんにちは')
  .then(res => console.log(res));

  console.log('こんばんは');
}
func();
```

```:出力
> おはよう
> こんばんは
> こんにちは
```

プログラム内で`Promise`の順番は上から２番目なのに、処理は最後にされています。



# コールスタックとキューとイベントループ

一体なぜ`Promise`のログは一番最後になったのでしょうか。

ここで登場するのが

- コールスタック
- キュー
- イベントループ

という3つの概念です。


## コールスタック

全ての関数は、呼び出されてコールスタックに入ります。
入った順に処理されて、処理が終わるとコールスタックから出ていきます。

![スクリーンショット 2020-08-21 22.51.42.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/b330ef5c-f4a6-d51a-2105-215e9650bd10.png)


## キュー

非同期処理も、まずはコールスタックに入ります。
ただし、非同期処理のコールバックはキューという別の場所に追加されます。

![スクリーンショット 2020-08-22 1.10.30.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/02ce14f7-8170-998e-43ce-a21630312d12.png)


## イベントループ

コールスタック内の全ての処理が終了した後で、ようやくキュー内の処理の出番がやってきます。

コールスタック内の全ての処理が終了した後、キューに処理が残っている場合は、イベントループがコールスタックにその処理を追加します。

![スクリーンショット 2020-08-22 1.15.11.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/6772d9dd-e581-bc60-801a-1e4ca81966b3.png)


このような流れになっているため、**一度キューに入る非同期処理は、同期処理よりも後に実行されることになります。**



# Async と Await

「書いた順番通りに実行されないなんて、読みにくいし気持ちわるい」

そう思う人のために、`Async` と `Await`という仕組みがJSに導入されました。

以下のように`Async` と `Await`を書き足すだけで、まるで同期処理のように順番通り出力されます。

```javascript
const func = async () => {
  console.log('おはよう');

  await Promise.resolve('こんにちは')
    .then(res => console.log(res));

  console.log('こんばんは');
}
func();
```

```:出力
> おはよう
> こんにちは
> こんばんは
```


## AwaitはAsync関数を一時停止する

一体なぜ上から順に`console.log()`出力がされたのでしょうか。

実は`await`はその時点でプログラムの実行を一旦止め、Promiseが解決されたら次に進むようにするキーワードです。

つまり、`Promise`が解決されるのを待っていてくれるのです。

`await`以下に記述された処理は、Promiseが解決されるまで実行されません。
したがって、上から順に処理を実行することができるわけです。


## QiitaAPIを今度は`await`で

先ほどは`then()`を使って`Promise`に返ってきた値を見ることができましたが、同じことが`await`でも可能です。
`await`は`Promise`値が返ってくるまで、待っていてくれるからです。

```javascript
(await fetch("https://qiita.com/api/v2/items")).json();
```
![スクリーンショット 2020-08-22 1.59.30.png](https://qiita-image-store.s3.ap-northeast-1.amazonaws.com/0/664546/71563341-6d30-8a0a-d25b-1e5865ca7f46.png)



# 【注意】非同期を非同期として使おう

`Async` と `Await`を使うことで、非同期処理を待って次の処理を行うことができるようになりました。

全部`Async` と `Await`で同期処理っぽく書いておけば、読みやすいし扱いやすいような気がしてきます。

しかし、濫用は禁物です。非同期の本来の良さを無駄にしてしまう可能性があります。


## 非同期を同期処理にしてしまった例

以下のコードを見てください。
全てのデータと、古いデータを比較して、新しいデータだけを表示する関数です。

```javascript
async function showNewData() {
  const allData = await fetchAllData();
  const oldData = await fetchOldData();
  showData(allData, oldData);
}
```

問題なのは、`allData`を問い合わせている間、この関数全体が止まってしまうことです。

非同期処理をする本来の目的は、**時間のかかる処理の完了を待つことなく、次の処理を行うこと**でした。
つまり、

- `allData`の問い合わせ と
- `oldData`の問い合わせ

は、同時に行えば効率が良いのに、`await`で処理を一旦止めてしまっているのです。


## Promise.all()

複数の非同期処理をまとめ、全て完了させた後にコールバックできるAPIが用意されています。
以下のように書くことができます。

```javascript
async function showNewData() {
  const [allData, oldData] = await Promise.all([fetchData(), fetchOldData()]);
  showData(allData, oldData);
}
```


# まとめ

以下を学習しました。

- `Promise`
- `then()` と `catch()`
- `async` と `await`
- コールスタックとキューとイベントループ

上記を知ることで、

- 「なぜ」非同期処理を行うのか
- 「どうやって」非同期処理が実現されているのか
- 「いつ」使うのが適切か

など、なんとなくではない意図した非同期処理が少しでもできるといいなと思います。


---

- **[【JavaScript】本日未明、[ if - else ]さんが死体で発見され...](https://qiita.com/impl_s/items/57aeeb2cb9b6773ae2fd)**




